### PR TITLE
Notifications for failed network requests

### PR DESCRIPTION
**Pull request by DivyanshiSingh**
Fixes: #13 
1) Created a reuseable Notification component to display the error messages.
2) Created an HttpApiCallError class to handle the errors.
3) Used try-catch in async functions.

### CODE REVIEW

**Comment by DivyanshiSingh** [comment]
This is my initial implementation, please review it @jennydaman

File: src/errors/HttpApiCallError.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+export default class HttpApiCallError extends Error {
+    constructor(message, statusCode) {
```
Comment (jennydaman) [inline src/errors/HttpApiCallError.js:2]: Please fix indentation here

File: src/components/Plugin/Plugin.js [hunk 2]
```diff
@@ -41,12 +44,15 @@ export class Plugin extends Component {
...
-      pluginData = await this.fetchPluginData();
-    } else {
-      ({ pluginData } = this.state);
+    try{
+      if (!this.state.pluginData) {
+        pluginData = await this.fetchPluginData();
+      } else {
+        ({ pluginData } = this.state);
+      }
+    }catch(errors){
```
Comment (jennydaman) [inline src/components/Plugin/Plugin.js:53]: Future work would be to handle `errors` specifically.

File: src/components/Plugin/Plugin.js [hunk 3]
```diff
@@ -41,12 +44,15 @@ export class Plugin extends Component {
...
+      this.showNotifications(new HttpApiCallError('Unable to fetch plugin data'));
     }
-
     if (this.isLoggedIn()) {
       this.fetchStarDataByPluginName(pluginData.name);
```
Comment (jennydaman) [inline src/components/Plugin/Plugin.js:57]: Put this in the `try` block too, it can fail.

File: src/components/Notifications/Notifications.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import React, { Component } from "react";
+import "./Notifications.css";
+class Notifications extends Component {
+  constructor(props) {
+    super(props);
+    this.deleteTime = 2000;
+  }
+  componentDidMount() {
+    this.timer = setInterval(() => {
+      console.log("here");
```
Comment (jennydaman) [inline src/components/Notifications/Notifications.js:11]: delete

File: src/components/Notifications/Notifications.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+      this.props.closeNotification();
+    }, this.deleteTime);
+    console.log(this.props.message, this.props);
```
Comment (jennydaman) [inline src/components/Notifications/Notifications.js:14]: and that

**Review by jennydaman — COMMENTED** [review]
This work is pretty amazing. A few coding things, like indentation and removal of print statements, have been commented upon.

Some feedback on style and notifications: The ChRIS_store_ui currently does not use many motion animations, so your notification feature stands out a bit. I think this could be alleviated by

- increasing the speed of the animation by 2-4x
- combine fade-in animation with slide-in animation
- change slide-in animation to move normal to the length of the notification box. i.e., have it slide in from the top instead of from the side.
- implementing an exit animation
- maybe have the exit animation be triggered by click, instead of on a timeout?

Error handling becomes a mess when you realize that there's a thousand ways something can go wrong for every way it can go right. Are you interested in adding the notification to more situations, or are you happy with what you have right now?

chris_store_ui_network_notification_1

**Comment by DivyanshiSingh** [comment]
Ok, @jennydaman I will do all the required changes. 
Thank you for your feedback.

**Comment by DivyanshiSingh** [comment]
> Some feedback on style and notifications: The ChRIS_store_ui currently does not use many motion animations, so your notification feature stands out a bit. I think this could be alleviated by
> * increasing the speed of the animation by 2-4x
   
> * combine fade-in animation with slide-in animation
> * change slide-in animation to move normal to the length of the notification box. i.e., have it slide in from the top instead of from the side.

 **I have increased the animation speed along with the fade-in animation**

> * implementing an exit animation
> * maybe have the exit animation be triggered by click, instead of on a timeout?

**I was planning to use the Alert component's props to give the functionality of closing notification on click, but as we are using an older version of patternfly-react , its not present there. I can implement it from scratch it will not be a very big change but we will have to rework once we change patternfly-react to a latest version.** How do you want me to proceed with this @jennydaman

**Comment by jennydaman** [comment]
@DivyanshiSingh you're right, notification-style element is a pretty advanced feature and I agree that it would be better to hold off on a thorough implementation until we finish the migration to Patternfly4.

File: src/errors/HttpApiCallError.js (hunk 1 above)
Comment (DivyanshiSingh) [reply src/errors/HttpApiCallError.js]: Resolved

File: src/components/Plugin/Plugin.js (hunk 3 above)
Comment (DivyanshiSingh) [reply src/components/Plugin/Plugin.js:57]: Handled the failure inside the function fetchStarDataByPluginName

File: src/components/Plugin/Plugin.js (hunk 2 above)
Comment (DivyanshiSingh) [reply src/components/Plugin/Plugin.js:53]: I think to handle the errors specifically I would need the structure of json response that we get. That will help me to design the HttpApiCallError according to the status code. Is this the right approach?

Comment (DivyanshiSingh) [reply src/components/Plugin/Plugin.js:53]: @jennydaman do we get structured error response from the API? I can use that to handle specific errors

Comment (jennydaman) [reply src/components/Plugin/Plugin.js:53]: Well, some responses are structured, but of course not every possibility can be covered.

In my original comment, I was originally thinking about handling errors we know of, like

- network timeout
- unexpected HTTP code (400, 500)
- cannot understand response

Comment (DivyanshiSingh) [reply src/components/Plugin/Plugin.js:53]: Ok got it. Will do the changes and let you know

**Comment by jennydaman** [comment]
@DivyanshiSingh the imports are now in conflict, hmm...

If you would like to forward-port this feature to `pf4` using Patternfly4 components, that would be something great to have.

**Review by jennydaman — APPROVED** [review]
@DivyanshiSingh I have resolved a merge conflict with #134 

However since https://github.com/FNNDSC/ChRIS_store_ui/pull/144 was merged, maybe we should close this one. Is that okay with you, or do you want to merge this too?

**Comment by DivyanshiSingh** [comment]
We have to close this.

**Pull request closed without merging**